doc/designs/evp_skey.md

```diff
@@ -0,0 +1,184 @@
+provider level. The exact set of those params is provider-specific.
+
+It should be possible to export/import the raw algorithm-specific
+keydata of an EVP_SKEY from a provider to another provider. Of course a
+provider could refuse the export operation if the key is unexportable.
+
+The EVP_SKEY object *can* have an algorithm association. It should be a
+provider-specific decision or maybe an information that could be associated to
+an EVP_SKEY any time later. Whether it starts with being algorithm specific or
+not depends on how it was created. If it's a result of obtaining a PKCS#11
+secret key, then it will most likely be algorithm specific from the start.
+
+We do not see why we would want to assign an algorithm immediately during
+the creation of an EVP_SKEY as result of derive() operation, if the operation
+just creates bunch of bytes that can be used as a key equally well for HMAC or
+AES encryption.
+
+Objects relationships
+---------------------
+
+We have EVP_CIPHER object associated with a particular provider. A EVP_SKEY
+object is derived for the particular EVP_CIPHER object. After that it is used
+within a EVP_CIPHER_CTX object associated with the same EVP_CIPHER object.
+
+The way to reuse the key within the other provider is through the export/import
+dance. As mentioned before, it may fail.
+
+Key management
+--------------
+
+The provider operating EVP_SKEY objects has to implement a separate key
+management to deal with opaque object. We already have MAC legacy key
+management and it looks suitable for dealing with EVP_SKEY objects.
+
+Allocation and freeing
+--------
+
+```C
+EVP_SKEY * EVP_SKEY_new(void);
+void EVP_SKEY_free(EVP_SKEY *key);
+```
+
+Importing and exporting parameters
+----------------------------------
+
+```C
+int EVP_SKEY_import(EVP_CIPHER *cipher, EVP_SKEY **pskey, OSSL_PARAM params[]);
+int EVP_SKEY_export(const EVP_SKEY *key, OSSL_PARAM **params,
+                    OSSL_CALLBACK *export_cb, void *export_cbarg);
+```
+
+A parameter `raw_bytes` can be used if we try to get a key in classical raw
+bytes implementation.
+
+To find out the parameters supported by a particular EVP_CIPHER object,
+we need a function
+
+```C
+const OSSL_PARAM *EVP_SKEY_fromdata_settable(EVP_CIPHER *cipher);
+```
+
+similar to `EVP_PKEY_fromdata_settable`
+
+API to derive a symmetric key
+-----------------------------
+
+The derived key can be algorithm-specific or algorithm-agnostic. To specify the
+algorithm binding, the params argument can be used.
+
+```C
+int EVP_PKEY_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey,
+                         OSSL_PARAM params[]);
+int EVP_KDF_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey,
+                        OSSL_PARAM params[]);
+```
+
+similar to `EVP_PKEY_derive/EVP_KDF_derive`
+
+Duplicating the EVP_SKEY object
+-------------------------------
+
+**Open question:** do we need it?
+
+```C
+EVP_SKEY * EVP_SKEY_dup(const EVP_SKEY * skey);
+```
```

Comment:
What is a hardware device supposed to do with it? Should it actually make a copy, and then have 2 copies in the hardware of the same key? I would at least wait with it until you have an actual use case for it.

Reply:
I will check whether I keep the current architecture as simple as it is now.

Reply:
Preferably an EVP_SKEY would be an immutable object inside the provider, so yeah, there is no point in having a dup function.

And yeah you need refcounting on the EVP level.

Reply:
IMHO the EVP_SKEY can be a handle to a hardware-backed key. E.g. for pkcs#11, is consists of a pkcs#11 reference (URI) and, later in the lifetime of the key, it gets a object handle. For cipher operations, the object handle is used for the cryptoki calls. If a key with the same URI is used in two different sessions, this can result in different object handles. (pkcs#11 folks, please correct me if I'm wrong). In such a case, duplicating the EVP_SKEY might duplicate the URI, but set the object handle to CK_INVALID_HANDLE.

For other hardware-backed keys, like s390 protected keys, duplication would also make sense, as the wrapped key can be copied as any other symmetric key as well.

Reply:
@holger-dengler within the lifetime of an application keys typically do not change handles, different application runs might get different handles though.

Reply:
@simo5 I agree, all pkcs#11 implementations I know use identical object handle values for a given object. But on the other hand, the pkcs#11 standard guaranties it only for the same session. AFAIK, it is not uncommon to use multiple sessions in a single process, so such use cases should in my opinion be aware of different object handle values for the same key object, if they are used in different sessions. 

E.g. pkcs#11 providers can use the `EVP_SKEY_dup()` function to reset the object handle, so that the duplicates can be used in different sessions.

Reply:
No, handles are valid across sessions, for example you can create a session key in one session and reference them in another.

If you duplicate a key then you do not have the same key, you have two distinct objects that just happen to carry the same bytes.

Those two objects will definitely have different handles.

An implementation may return different handles for the same object I guess, but it would not be well behaved if it does that, and would still need to allow the same handle to be honored in any session.

Reply:
> No, handles are valid across sessions, for example you can create a session key in one session and reference them in another.

Hm, I understand the pkcs#11 standard [1] differently:
```
When an object is created or found on a token by an application, Cryptoki assigns it an object
handle for that application’s sessions to use to access it.  A particular object on a token does
not necessarily have a handle which is fixed for the lifetime of the object; however, if a particular
session can use a particular handle to access a particular object, then that session will continue
to be able to use that handle to access that object as long as the session continues to exist,
the object continues to exist, and the object continues to be accessible to the session.
```
In my understanding that means, that the value of the object handle is only guarantied within a session, but maybe the fault is on my side.

> If you duplicate a key then you do not have the same key, you have two distinct objects that just happen to carry the same bytes.
You're right. So just reset the object handle is not enough in case of pkcs#11, the object should be copied (if it is copy-able). With this, you can later change flags of the duplicate object without modifying the original one.

Anyhow, if a provider can not duplicate the referenced key of an EVP_SKEY (e.g. if the hardware does not support key duplication), duplicate() should return with an error.

[1] https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/os/pkcs11-base-v3.0-os.html#_Toc29976544 , in section CK_OBJECT_HANDLE

Reply:
@simo5 thanks for clarification. I've to think about this a bit now... ;)